src/eos-updater-poll.c

```diff
@@ -203,6 +203,192 @@ get_fetchers (SourcesConfig *config,
+}
+
+static GPtrArray *
+get_finders (SourcesConfig          *config,
+             GMainContext           *context,
+             OstreeRepoFinderAvahi **out_finder_avahi)
+{
+  g_autoptr(OstreeRepoFinderAvahi) finder_avahi = NULL;
+  g_autoptr(GPtrArray) finders = g_ptr_array_new_full (config->download_order->len,
+                                                       object_unref0);
+  g_autoptr(GError) local_error = NULL;
+  gsize i;
+
+  g_assert (config->download_order->len > 0);
+
+  for (i = 0; i < config->download_order->len; i++)
+    {
+      switch (g_array_index (config->download_order, EosUpdaterDownloadSource, i))
+        {
+        case EOS_UPDATER_DOWNLOAD_MAIN:
+          g_ptr_array_add (finders, ostree_repo_finder_config_new ());
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_LAN:
+          finder_avahi = ostree_repo_finder_avahi_new (context);
+          g_ptr_array_add (finders, g_object_ref (finder_avahi));
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_VOLUME:
+          /* TODO: How to make this one testable? */
+          g_ptr_array_add (finders, ostree_repo_finder_mount_new (NULL));
+          break;
+
+        default:
+          g_assert_not_reached ();
+        }
+    }
+
+  g_ptr_array_add (finders, NULL);  /* NULL terminator */
+
+  /* TODO: Stop this at some point; think of a better way to store it and
+   * control its lifecycle. */
+  if (finder_avahi != NULL)
+    ostree_repo_finder_avahi_start (OSTREE_REPO_FINDER_AVAHI (finder_avahi),
+                                    &local_error);
+
+  if (local_error != NULL)
+    {
+      g_warning ("Avahi finder failed; removing it: %s", local_error->message);
+      g_ptr_array_remove (finders, finder_avahi);
+      g_clear_object (&finder_avahi);
+      g_clear_error (&local_error);
+    }
+
+  if (out_finder_avahi != NULL)
+    *out_finder_avahi = g_steal_pointer (&finder_avahi);
+
+  return g_steal_pointer (&finders);
+}
+
+static void
+async_result_cb (GObject      *source_object,
+                 GAsyncResult *result,
+                 gpointer      user_data)
+{
+  GAsyncResult **result_out = user_data;
+
+  *result_out = g_object_ref (result);
+}
+
+static EosUpdateInfo *
+metadata_fetch_new (OstreeRepo    *repo,
+                    SourcesConfig *config,
+                    GMainContext  *context,
+                    GCancellable  *cancellable,
+                    GError       **error)
+{
+  g_auto(OstreeRepoFinderResultv) results = NULL;
+  g_autoptr(EosUpdateInfo) info = NULL;
+  g_autofree gchar *checksum = NULL;
+  g_autoptr(GVariant) commit = NULL;
+  g_autofree gchar *booted_refspec = NULL, *new_refspec = NULL;
+  g_autoptr(OstreeCollectionRef) booted_collection_ref = NULL, new_collection_ref = NULL;
+  const gchar *upgrade_refspec;
+  const OstreeCollectionRef *upgrade_collection_ref;
+  g_autoptr(GPtrArray) finders = NULL;  /* (element-type OstreeRepoFinder) */
+  g_autoptr(OstreeRepoFinderAvahi) finder_avahi = NULL;
```

Comment:
It looks like this is unused

Reply:
Oops, I need to stop it searching once polling is complete.

Reply:
Fixup pushed.